doc/tools/docgram/doc_grammar.ml

```diff
@@ -309,6 +311,11 @@ let rec output_prodn = function
             | "{|" -> "%{%|"
             | "`{" -> "`%{"
             | "@{" -> "@%{"
+            | "|-" -> "%|-"
+            | "|->" -> "%|->"
+            | "||" -> "%||"
+            | "|||" -> "%|||"
+            | "||||" -> "%||||"
```

Comment:
These are not really syntax, but rather hacks to make the parser resilient to Coq's lexer (see for example https://github.com/coq/coq/issues/13184)

In ssr scripts you can find `=> [||| x]` and expect it to be parsed it as `=> [ | | | x ]` but Coq's lexer/parser is weak and if someone declares a notation like `"x ||| y"` then Coq would not parse `|||` as 3 separate tokens but as one, breaking the parsing rule.  The same for `[|->]` to be parsed as `[ | -> ]`.
As you can expect the list of hacks is not exhaustive and only covers the most common conflicts.

Reply:
I see.  I'll leave these out of the doc.

"||" is used in ltac and in a number of notations.  There are similar tricks used for other non-ident keywords.  Thinking out loud, perhaps we could change the infrastructure so that multi-character non-identifier keywords such as "||" are processed as multiple tokens, one per character, e.g. "|" "|"?  Code for the `Notation` command would need to change.  But this would be simpler to understand and work with for developers and users.

I notice that mtac2 has a number of tokens like "[m:" and they even have "m:{", so we'd want to consider how to handle those cases.  And it could make some sequences unparsable with our single-token lookahead.  But maybe that's worth trying to change, too.